File: src/leanResponse.ts
// TOKEN ECONOMY — every MCP tool response is shaped here before it leaves the server.
//
// WHY THIS EXISTS (measured, 2026-07-09): a diagnostic tool's *code* is free, but its *output* is not.
// A tool result lands in the caller's conversation transcript, and the Anthropic API is stateless — the
// whole transcript is re-sent on EVERY subsequent turn. So a 12 KB JSON blob is not paid once; it is
// paid again on every later turn until compaction. In a 1.2M-token marathon session that single blob
// silently costs more than the entire analysis it reports.
//
// THE CONTRACT: a diagnostic tool must RETURN THE ANSWER, not the raw data to derive it. The server
// already computes the verdict, the ranking, and the remediation — the caller should get exactly that,
// bounded, and nothing else. Arrays are truncated to their significant head; `coverage` collapses to
// one honest line; the fields listed in DROP_KEYS (derivation detail) are removed. Truncation is
// ALWAYS disclosed (never a silent cut) and the full payload stays one `verbosity:"full"` away.
//
// NESTED OBJECTS ARE KEPT. They were flattened away until 2026-07-27, on the premise that nesting
// meant derivation — false for any tool whose answer is a structured object, and it silently deleted
// 87% of get_window_budget's leaves, 84% of get_burn_status's, and the authoritative
// usageWindows.fiveHourPct from get_account_status. Depth cannot decide the question: within one
// window object `pctConsumed` (answer) and `breakdown` (derivation) are peers, both all-scalar. So
// the split is DECLARED in DROP_KEYS and nowhere else, and the ceiling below is what bounds cost.
//
// The ceiling is enforced at the end as a hard backstop: if a shaped payload still exceeds the budget,
// it is progressively degraded (arrays shrink, then long strings clip) until it fits. A tool can never
// blow the caller's context, no matter what the underlying data looks like.

export type Verbosity = 'summary' | 'full'

/** ~4 chars/token is the standard rough estimator; we only need the order of magnitude for a cap. */
const CHARS_PER_TOKEN = 4
/** Default ceiling for a summary response (~1.2k tokens). Chosen so a dozen calls stay under ~15k. */
const DEFAULT_MAX_TOKENS = 1200
const MAX_ROWS = 5
const MAX_STR = 400
/** Elements kept from an array nested inside a row (a disclosure marker is appended when it cuts). */
const MAX_NESTED_ROWS = 3
/** Recursion guard ONLY — a cycle/pathology backstop, never the semantic filter. Measured across the
 *  shipped tools the deepest real answer is depth 3 (`accounts[].budget.fiveHour.*`), so 4 clears
 *  every tool with room to spare. What counts as derivation is decided by DROP_KEYS, not by depth:
 *  `fiveHour.pctConsumed` (answer) and `fiveHour.breakdown` (derivation) sit at the SAME depth and
 *  are both all-scalar, so no structural rule can separate them. */
const MAX_DEPTH = 4

export interface LeanOptions {
  verbosity?: Verbosity
  maxTokens?: number
}

function approxTokens(v: unknown): number {
  return Math.ceil(JSON.stringify(v ?? '').length / CHARS_PER_TOKEN)
}

function clipStr(s: string, max = MAX_STR): string {
  return s.length <= max ? s : `${s.slice(0, max)}… (+${s.length - max} chars)`
}

/** Collapse a verbose `coverage` object into ONE honest line — the scan scope must never be hidden,
 *  but it also never needs 8 fields to be understood. */
function coverageLine(cov: unknown): string | undefined {
  if (!cov || typeof cov !== 'object') return undefined
  const c = cov as Record<string, unknown>
  if (typeof c.note === 'string' && c.note) return clipStr(c.note, 240)
  if (c.complete === true) return 'complete scan'
  return undefined
}

/**
 * Row-level shaping: keep the fields that carry the ANSWER, drop the ones that carry the derivation.
 *
 * This set is the ONE place the answer/derivation split is declared, and it is deliberately a
 * DENY-list rather than an allow-list: a missing entry costs a few extra (ceiling-capped, visible)
 * tokens, whereas a missing allow-list entry would silently delete a tool's answer — which is exactly
 * the defect this shaper shipped with. Keep-by-default fails safe; drop-by-default fails silent.
 *
 * `breakdown` — the per-bucket {input, output, cacheRead, cacheCreation, …} split that hangs off
 * every window/rate object. Measured across the live payloads of get_window_budget and
 * get_burn_status it is derivation in 100% of its occurrences and never an answer; the totals beside
 * it (consumedTokens / consumedCostUsd / consumedBillableWeighted) carry the verdict.
 *
 * `remediation` is NOT here, though it once was: it is genuinely returned (cacheBreakTimeline.ts,
 * mcpServer.ts:2436) and four tool descriptions advertise it as part of the answer ("an ordered
 * `remediation` list ranked by what actually dominates the cost"). Dropping it made those tools
 * promise a fix hint they never delivered.
 *
 * The machine-identity keys below (culpritId/actorId/bodyRef/ttlTier) stay dropped because each has
 * a human-readable sibling that survives — culpritSummary, actor, cause.
 */
const DROP_KEYS = new Set(['rawDiffSummary', 'culpritId', 'actorId', 'ttlTier', 'bodyRef', 'ref', 'breakdown'])

function shapeRow(row: unknown, depth = 0): unknown {
  if (row === null || typeof row !== 'object') return typeof row === 'string' ? clipStr(row) : row
  // NB: `.map(shapeRow)` would hand the array INDEX to `depth` — always pass it explicitly.
  if (Array.isArray(row)) return row.slice(0, MAX_ROWS).map(r => shapeRow(r, depth))
  const out: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(row as Record<string, unknown>)) {
    if (DROP_KEYS.has(k)) continue
    if (v === null || v === undefined) continue
    if (typeof v === 'string') { out[k] = clipStr(v, 160); continue }
    if (typeof v === 'number' || typeof v === 'boolean') { out[k] = v; continue }
    if (Array.isArray(v)) {
      const kept = v.slice(0, MAX_NESTED_ROWS).map(r => shapeRow(r, depth + 1))
      // Disclose the cut — the file's contract is that truncation is NEVER silent.
      out[k] = v.length > MAX_NESTED_ROWS
        ? [...kept, `… +${v.length - MAX_NESTED_ROWS} more — use verbosity:"full"`]
        : kept
      continue
    }
    // Nested object: KEEP it (recursing), because for a structured tool the nested object IS the
    // answer — accounts[].budget.fiveHour.pctConsumed, usageWindows.fiveHourPct, risks[].evidence.
    // DROP_KEYS above is what removes derivation; depth is only a pathology guard.
    if (depth + 1 >= MAX_DEPTH) {
      out[k] = `… ${Object.keys(v as object).length} field(s) elided at depth ${MAX_DEPTH} — use verbosity:"full"`
      continue
    }
    out[k] = shapeRow(v, depth + 1)
  }
  return out
}

/** Truncate an array to its significant head, DISCLOSING what was dropped (never a silent cut). */
function headOf(arr: unknown[], limit: number, label: string): { rows: unknown[]; note?: string } {
  // `.map(shapeRow)` would hand the array INDEX to `depth`, so element N would be shaped as if it sat
  // N levels deep and its answer would elide at the depth guard — the top-level rows of every
  // array-shaped tool (get_window_budget's accounts, check_burn_risk's risks) start at depth 0.
  if (arr.length <= limit) return { rows: arr.map(r => shapeRow(r, 0)) }
  return {
    rows: arr.slice(0, limit).map(r => shapeRow(r, 0)),
    note: `showing top ${limit} of ${arr.length} ${label} — call with verbosity:"full" for all`,
  }
}

/** Generic shaper: verdict/notes first, then the significant head of the biggest ranked array. */
function shapeGeneric(result: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  const notes: string[] = []

  // 1. Scalars + the verdict-bearing fields lead the payload — this IS the answer.
  for (const [k, v] of Object.entries(result)) {
    if (k === 'coverage') continue
    if (typeof v === 'string') { out[k] = clipStr(v, k === 'verdict' ? 600 : MAX_STR); continue }
    if (typeof v === 'number' || typeof v === 'boolean') out[k] = v
  }

  // 2. Arrays → truncated heads, each disclosing its own truncation.
  for (const [k, v] of Object.entries(result)) {
    if (!Array.isArray(v) || v.length === 0) continue
    const { rows, note } = headOf(v, MAX_ROWS, k)
    out[k] = rows
    if (note) notes.push(`${k}: ${note}`)
  }

  // 3. Nested objects: keep only their scalar summary (one level), never the whole tree.
  for (const [k, v] of Object.entries(result)) {
    if (k === 'coverage' || Array.isArray(v) || v === null || typeof v !== 'object') continue
    const shaped = shapeRow(v)
    if (shaped && Object.keys(shaped as object).length) out[k] = shaped
  }

  const cov = coverageLine(result.coverage)
  if (cov) out.coverage = cov
  if (notes.length) out._truncated = notes
  return out
}

/** Deepest object-nesting level in a value (a flat object is 1; arrays are transparent). */
function objectDepth(v: unknown): number {
  if (v === null || typeof v !== 'object') return 0
  if (Array.isArray(v)) return v.reduce<number>((m, x) => Math.max(m, objectDepth(x)), 0)
  let deepest = 0
  for (const x of Object.values(v as Record<string, unknown>)) deepest = Math.max(deepest, objectDepth(x))
  return deepest + 1
}

/** Replace every nested object sitting at or below `maxLevel` with a disclosed marker. Root is
 *  level 0, so maxLevel=2 keeps the root's direct children and elides their children. */
function pruneBelow(v: unknown, maxLevel: number, level = 0): unknown {
  if (v === null || typeof v !== 'object') return v
  if (Array.isArray(v)) return v.map(x => pruneBelow(x, maxLevel, level))
  const out: Record<string, unknown> = {}
  for (const [k, x] of Object.entries(v as Record<string, unknown>)) {
    if (x !== null && typeof x === 'object' && !Array.isArray(x)) {
      out[k] = level + 1 >= maxLevel
        ? `… ${Object.keys(x as object).length} field(s) elided — use verbosity:"full"`
        : pruneBelow(x, maxLevel, level + 1)
      continue
    }
    out[k] = pruneBelow(x, maxLevel, level)
  }
  return out
}

/** Clip long strings at EVERY level, not just the root — a deep payload's bulk is rarely top-level. */
function clipDeep(v: unknown, max: number): unknown {
  if (typeof v === 'string') return clipStr(v, max)
  if (v === null || typeof v !== 'object') return v
  if (Array.isArray(v)) return v.map(x => clipDeep(x, max))
  const out: Record<string, unknown> = {}
  for (const [k, x] of Object.entries(v as Record<string, unknown>)) out[k] = clipDeep(x, max)
  return out
}

/** Keep the first `maxKeys` entries of every object, disclosing the count dropped. The last resort:
 *  a payload can be too big by being WIDE (many sibling keys) rather than deep or array-heavy, and
 *  by the time this runs the deep values are already elision markers, so little is lost. */
function narrowTo(v: unknown, maxKeys: number): unknown {
  if (v === null || typeof v !== 'object') return v
  if (Array.isArray(v)) return v.map(x => narrowTo(x, maxKeys))
  const entries = Object.entries(v as Record<string, unknown>)
  const out: Record<string, unknown> = {}
  for (const [k, x] of entries.slice(0, maxKeys)) out[k] = narrowTo(x, maxKeys)
  if (entries.length > maxKeys) out._elidedKeys = `+${entries.length - maxKeys} more field(s) — use verbosity:"full"`
  return out
}

/** Hard backstop: degrade until the payload fits the ceiling, disclosing every step that changes
 *  anything. Arrays shrink first (most compressible), then object DEPTH is pruned, then object WIDTH,
 *  then long strings clip at every level.
 *
 *  Depth and width both exist because nested objects are now kept. Without depth, a tool whose bulk is
 *  one deep object is unrecoverable once arrays are down to a single row; without width, a wide flat
 *  object floors out above the ceiling no matter how deeply pruned — which made the "can never blow
 *  the caller's context" promise false until a test proved it. A tool can never blow the caller's
 *  context, whatever the data looks like. */
function enforceCeiling(shaped: Record<string, unknown>, maxTokens: number): Record<string, unknown> {
  let out = shaped
  // ONE ceiling note, rewritten in place — never one per iteration. Each loop below steps its limit
  // down repeatedly (arrays 3→2→1, width 16→8→4→2→1), and appending a note per step was wrong twice
  // over: the intermediate lines are FALSE by the time the loop settles ("arrays reduced to 3" when
  // they ended at 1), and ~15 of them at ~110 chars each is ~190 tokens of disclosure that the
  // ceiling then cannot absorb. Measured: a payload degraded against a 60-token ceiling came out at
  // 203 tokens, of which 7 notes WERE essentially the whole payload — the function breaching its own
  // promise with the text explaining that it had not. One replaceable note is bounded and true.
  const CEILING_NOTE = 'payload exceeded'
  let applied = new Map<string, string>()

  /** Apply a degradation; keep it only if it actually shrank the payload — disclosure INCLUDED, so a
   *  stage that cannot pay for its own note is rejected and the next iteration simply degrades
   *  harder. No-op stages (a payload with no arrays, say) never emit a note for work they did not do.
   *  Notes from EARLIER phases (the shaper's own coverage note) are preserved; only the ceiling's own
   *  line is replaced. */
  const stage = (candidate: Record<string, unknown>, kind: string, detail: string): boolean => {
    const merged = new Map(applied).set(kind, detail)
    const note = `${CEILING_NOTE} ~${maxTokens} tokens — `
      + [...merged].map(([k, d]) => `${k} ${d}`).join(', ')
      + '; use verbosity:"full"'
    const kept = (Array.isArray(out._truncated) ? (out._truncated as string[]) : [])
      .filter(n => !n.startsWith(CEILING_NOTE))
    const next: Record<string, unknown> = { ...candidate, _truncated: [...kept, note] }
    if (approxTokens(next) >= approxTokens(out)) return false
    applied = merged
    out = next
    return true
  }

  for (let limit = MAX_ROWS; limit >= 1 && approxTokens(out) > maxTokens; limit--) {
    const next: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(out)) next[k] = Array.isArray(v) ? v.slice(0, limit) : v
    stage(next, 'arrays', `→ ${limit} row(s)`)
  }

  for (let level = objectDepth(out) - 1; level >= 1 && approxTokens(out) > maxTokens; level--) {
    stage(pruneBelow(out, level) as Record<string, unknown>, 'nesting', `→ elided below level ${level}`)
  }

  for (let keys = 16; keys >= 1 && approxTokens(out) > maxTokens; keys = Math.floor(keys / 2)) {
    stage(narrowTo(out, keys) as Record<string, unknown>, 'width', `→ ${keys} field(s)`)
  }

  if (approxTokens(out) > maxTokens) {
    stage(clipDeep(out, 120) as Record<string, unknown>, 'strings', '→ clipped to 120 chars')
  }
  return out
}

/**
 * Shape a tool result for the caller's context.
 *
 * `verbosity:"full"` returns the payload untouched (the escape hatch for a genuine deep drill).
 * Otherwise the result is reduced to: scalars + verdict + the head of each ranked array + a one-line
 * coverage note, under a hard token ceiling, with every truncation disclosed in `_truncated`.
 *
 * Non-object results (a pre-rendered `{format,text}` table, a string) pass through unchanged except for
 * the ceiling — those are already the tool's own concise rendering.
 */
export function leanify(result: unknown, opts: LeanOptions = {}): unknown {
  const verbosity: Verbosity = opts.verbosity === 'full' ? 'full' : 'summary'
  if (verbosity === 'full') return result
  const maxTokens = opts.maxTokens ?? DEFAULT_MAX_TOKENS
  if (result === null || result === undefined) return result

  // A pre-rendered text payload ({format,text}) is the tool's own compact rendering — only cap it.
  if (typeof result === 'object' && !Array.isArray(result) && typeof (result as { text?: unknown }).text === 'string') {
    const r = result as { format?: unknown; text: string }
    const budget = maxTokens * CHARS_PER_TOKEN
    return r.text.length <= budget
      ? result
      : { ...r, text: `${r.text.slice(0, budget)}\n… truncated (${r.text.length} chars) — use verbosity:"full"` }
  }
  if (typeof result !== 'object') return result
  if (Array.isArray(result)) {
    const { rows, note } = headOf(result, MAX_ROWS, 'rows')
    return enforceCeiling(note ? { rows, _truncated: [note] } : { rows }, maxTokens)
  }
  return enforceCeiling(shapeGeneric(result as Record<string, unknown>), maxTokens)
}

File: rust-core/crates/agentlens-core/src/lean_response.rs
//! Port of src/leanResponse.ts (TRDD-DMWOBWFH P4x.2c) — the single choke point every MCP tool
//! response passes through before it leaves the server.
//!
//! WHY IT MATTERS HERE: a tool result lands in the CALLER's transcript, and the API is stateless —
//! the whole transcript is re-sent on every later turn. A 12 KB blob is not paid once, it is paid
//! again every turn until compaction. The Rust core served 9 tools raw before this landed, which
//! made every one of them a different (and more expensive) wire shape than the TS it replaces.
//!
//! Two rules the TS states and this port must not quietly "improve":
//!   - the answer/derivation split is DECLARED in `DROP_KEYS` and nowhere else. It is a DENY-list on
//!     purpose: a missing deny entry costs a few visible tokens, a missing allow entry would
//!     silently delete a tool's answer. Keep-by-default fails safe; drop-by-default fails silent.
//!   - truncation is NEVER silent. Every cut discloses itself, and the full payload stays one
//!     `verbosity: "full"` away.
//!
//! JS-ISM: every length and slice here is UTF-16 code units (`String.prototype.length`/`.slice`),
//! never bytes and never chars. The elision markers themselves contain a `…` (1 UTF-16 unit, 3
//! UTF-8 bytes), so measuring in bytes would drift the ceiling on exactly the payloads the ceiling
//! is meant to bound.

use indexmap::IndexMap;
use serde_json::{Map, Value};

use crate::summarize::helpers::{js_slice, utf16_len};

const CHARS_PER_TOKEN: f64 = 4.0;
const DEFAULT_MAX_TOKENS: f64 = 1200.0;
const MAX_ROWS: usize = 5;
const MAX_STR: usize = 400;
const MAX_NESTED_ROWS: usize = 3;
/// Recursion guard ONLY — a pathology backstop, never the semantic filter. `fiveHour.pctConsumed`
/// (answer) and `fiveHour.breakdown` (derivation) sit at the SAME depth and are both all-scalar, so
/// no structural rule can separate them; `DROP_KEYS` does.
const MAX_DEPTH: usize = 4;

/// The machine-identity keys each have a human-readable sibling that survives (culpritSummary,
/// actor, cause). `remediation` is deliberately NOT here — it is genuinely part of four tools'
/// advertised answer, and dropping it made them promise a fix hint they never delivered.
const DROP_KEYS: [&str; 7] = ["rawDiffSummary", "culpritId", "actorId", "ttlTier", "bodyRef", "ref", "breakdown"];

/// `Math.ceil(JSON.stringify(v ?? '').length / 4)`. The `?? ''` matters: a null payload measures as
/// `""` (2 units → 1 token), NOT as the 4 units of `null`.
fn approx_tokens(v: &Value) -> f64 {
    let s = if v.is_null() { "\"\"".to_owned() } else { v.to_string() };
    (utf16_len(&s) as f64 / CHARS_PER_TOKEN).ceil()
}

fn clip_str(s: &str, max: usize) -> String {
    let n = utf16_len(s);
    if n <= max {
        s.to_owned()
    } else {
        format!("{}… (+{} chars)", js_slice(s, max), n - max)
    }
}

/// Collapse a verbose `coverage` object into ONE honest line — the scan scope must never be hidden,
/// but it also never needs 8 fields to be understood.
fn coverage_line(cov: Option<&Value>) -> Option<String> {
    let c = cov?.as_object()?;
    if let Some(note) = c.get("note").and_then(Value::as_str).filter(|s| !s.is_empty()) {
        return Some(clip_str(note, 240));
    }
    if c.get("complete") == Some(&Value::Bool(true)) {
        return Some("complete scan".to_owned());
    }
    None
}

/// Row-level shaping: keep the fields carrying the ANSWER, drop the ones carrying the derivation.
///
/// `depth` is passed EXPLICITLY at every call site. The TS carries a comment about this twice
/// because `.map(shapeRow)` hands JS the array INDEX as `depth`, so element N would be shaped as if
/// it sat N levels deep and its answer would vanish at the depth guard. Rust closures do not have
/// that hazard, but the depth values must still match the TS exactly for the markers to agree.
fn shape_row(row: &Value, depth: usize) -> Value {
    match row {
        Value::String(s) => Value::String(clip_str(s, MAX_STR)),
        Value::Array(a) => Value::Array(a.iter().take(MAX_ROWS).map(|r| shape_row(r, depth)).collect()),
        Value::Object(o) => {
            let mut out = Map::new();
            for (k, v) in o {
                if DROP_KEYS.contains(&k.as_str()) || v.is_null() {
                    continue;
                }
                match v {
                    Value::String(s) => {
                        out.insert(k.clone(), Value::String(clip_str(s, 160)));
                    }
                    Value::Number(_) | Value::Bool(_) => {
                        out.insert(k.clone(), v.clone());
                    }
                    Value::Array(a) => {
                        let mut kept: Vec<Value> = a.iter().take(MAX_NESTED_ROWS).map(|r| shape_row(r, depth + 1)).collect();
                        // Disclose the cut — truncation is NEVER silent.
                        if a.len() > MAX_NESTED_ROWS {
                            kept.push(Value::String(format!(
                                "… +{} more — use verbosity:\"full\"",
                                a.len() - MAX_NESTED_ROWS
                            )));
                        }
                        out.insert(k.clone(), Value::Array(kept));
                    }
                    // Nested object: KEPT (recursing) — for a structured tool the nested object IS
                    // the answer (accounts[].budget.fiveHour.pctConsumed, risks[].evidence).
                    _ => {
                        if depth + 1 >= MAX_DEPTH {
                            out.insert(k.clone(), Value::String(format!(
                                "… {} field(s) elided at depth {MAX_DEPTH} — use verbosity:\"full\"",
                                v.as_object().map_or(0, Map::len)
                            )));
                        } else {
                            out.insert(k.clone(), shape_row(v, depth + 1));
                        }
                    }
                }
            }
            Value::Object(out)
        }
        _ => row.clone(),
    }
}

/// Truncate an array to its significant head, DISCLOSING what was dropped.
fn head_of(arr: &[Value], limit: usize, label: &str) -> (Vec<Value>, Option<String>) {
    let rows: Vec<Value> = arr.iter().take(limit).map(|r| shape_row(r, 0)).collect();
    if arr.len() <= limit {
        (rows, None)
    } else {
        (rows, Some(format!("showing top {limit} of {} {label} — call with verbosity:\"full\" for all", arr.len())))
    }
}

/// Generic shaper: verdict/notes first, then the significant head of each ranked array, then the
/// nested objects. The THREE PASSES are the wire's key ORDER — scalars, arrays, objects, coverage,
/// `_truncated` — not the input's order, so they cannot be merged into one loop.
fn shape_generic(result: &Map<String, Value>) -> Map<String, Value> {
    let mut out = Map::new();
    let mut notes: Vec<Value> = Vec::new();

    // 1. Scalars + the verdict-bearing fields lead the payload — this IS the answer.
    for (k, v) in result {
        if k == "coverage" {
            continue;
        }
        match v {
            Value::String(s) => {
                out.insert(k.clone(), Value::String(clip_str(s, if k == "verdict" { 600 } else { MAX_STR })));
            }
            Value::Number(_) | Value::Bool(_) => {
                out.insert(k.clone(), v.clone());
            }
            _ => {}
        }
    }

    // 2. Arrays → truncated heads, each disclosing its own truncation.
    for (k, v) in result {
        let Some(a) = v.as_array().filter(|a| !a.is_empty()) else { continue };
        let (rows, note) = head_of(a, MAX_ROWS, k);
        out.insert(k.clone(), Value::Array(rows));
        if let Some(n) = note {
            notes.push(Value::String(format!("{k}: {n}")));
        }
    }

    // 3. Nested objects.
    for (k, v) in result {
        if k == "coverage" || !v.is_object() {
            continue;
        }
        let shaped = shape_row(v, 0);
        if shaped.as_object().is_some_and(|o| !o.is_empty()) {
            out.insert(k.clone(), shaped);
        }
    }

    if let Some(cov) = coverage_line(result.get("coverage")) {
        out.insert("coverage".into(), Value::String(cov));
    }
    if !notes.is_empty() {
        out.insert("_truncated".into(), Value::Array(notes));
    }
    out
}

/// Deepest object-nesting level (a flat object is 1; arrays are transparent).
fn object_depth(v: &Value) -> usize {
    match v {
        Value::Array(a) => a.iter().map(object_depth).max().unwrap_or(0),
        Value::Object(o) => o.values().map(object_depth).max().unwrap_or(0) + 1,
        _ => 0,
    }
}

/// Replace every nested object at or below `max_level` with a disclosed marker. Root is level 0, so
/// `max_level = 2` keeps the root's direct children and elides THEIR children.
fn prune_below(v: &Value, max_level: usize, level: usize) -> Value {
    match v {
        Value::Array(a) => Value::Array(a.iter().map(|x| prune_below(x, max_level, level)).collect()),
        Value::Object(o) => {
            let mut out = Map::new();
            for (k, x) in o {
                if x.is_object() {
                    out.insert(
                        k.clone(),
                        if level + 1 >= max_level {
                            Value::String(format!(
                                "… {} field(s) elided — use verbosity:\"full\"",
                                x.as_object().map_or(0, Map::len)
                            ))
                        } else {
                            prune_below(x, max_level, level + 1)
                        },
                    );
                } else {
                    out.insert(k.clone(), prune_below(x, max_level, level));
                }
            }
            Value::Object(out)
        }
        _ => v.clone(),
    }
}

/// Clip long strings at EVERY level — a deep payload's bulk is rarely top-level.
fn clip_deep(v: &Value, max: usize) -> Value {
    match v {
        Value::String(s) => Value::String(clip_str(s, max)),
        Value::Array(a) => Value::Array(a.iter().map(|x| clip_deep(x, max)).collect()),
        Value::Object(o) => Value::Object(o.iter().map(|(k, x)| (k.clone(), clip_deep(x, max))).collect()),
        _ => v.clone(),
    }
}

/// Keep the first `max_keys` entries of every object, disclosing the count dropped. The last resort:
/// a payload can be too big by being WIDE rather than deep or array-heavy, and by the time this runs
/// the deep values are already elision markers, so little is lost.
fn narrow_to(v: &Value, max_keys: usize) -> Value {
    match v {
        Value::Array(a) => Value::Array(a.iter().map(|x| narrow_to(x, max_keys)).collect()),
        Value::Object(o) => {
            let mut out = Map::new();
            for (k, x) in o.iter().take(max_keys) {
                out.insert(k.clone(), narrow_to(x, max_keys));
            }
            if o.len() > max_keys {
                out.insert(
                    "_elidedKeys".into(),
                    Value::String(format!("+{} more field(s) — use verbosity:\"full\"", o.len() - max_keys)),
                );
            }
            Value::Object(out)
        }
        _ => v.clone(),
    }
}

/// Hard backstop: degrade until the payload fits, disclosing every step that changes anything.
/// Arrays shrink first (most compressible), then object DEPTH, then object WIDTH, then strings.
///
/// ONE ceiling note, REWRITTEN IN PLACE — never one per iteration. The TS records why twice over:
/// the intermediate lines are FALSE by the time a loop settles ("arrays reduced to 3" when they
/// ended at 1), and ~15 of them at ~110 chars is ~190 tokens of disclosure the ceiling then cannot
/// absorb — a payload degraded against a 60-token ceiling came out at 203 tokens, of which 7 notes
/// WERE essentially the whole payload: the function breaching its own promise with the text
/// explaining that it had not.
fn enforce_ceiling(shaped: Map<String, Value>, max_tokens: f64) -> Map<String, Value> {
    const CEILING_NOTE: &str = "payload exceeded";
    let mut out = Value::Object(shaped);
    // IndexMap, not HashMap: the note lists the applied stages in the order they were applied, and
    // re-applying a stage must keep its ORIGINAL position (JS `Map.set` on an existing key does).
    let mut applied: IndexMap<String, String> = IndexMap::new();

    // Apply a degradation; keep it only if it actually shrank the payload — DISCLOSURE INCLUDED, so
    // a stage that cannot pay for its own note is rejected and the next iteration degrades harder.
    // No-op stages (a payload with no arrays) never emit a note for work they did not do.
    let mut stage = |candidate: Value, kind: &str, detail: String, out: &mut Value| -> bool {
        let mut merged = applied.clone();
        merged.insert(kind.to_owned(), detail);
        let note = format!(
            "{CEILING_NOTE} ~{} tokens — {}; use verbosity:\"full\"",
            crate::summarize::helpers::fmt_js_num(max_tokens),
            merged.iter().map(|(k, d)| format!("{k} {d}")).collect::<Vec<_>>().join(", ")
        );
        // Notes from EARLIER phases (the shaper's own coverage note) are preserved; only the
        // ceiling's own line is replaced.
        let mut kept: Vec<Value> = out
            .get("_truncated")
            .and_then(Value::as_array)
            .map(|a| a.iter().filter(|n| !n.as_str().is_some_and(|s| s.starts_with(CEILING_NOTE))).cloned().collect())
            .unwrap_or_default();
        kept.push(Value::String(note));
        let mut next = candidate.as_object().cloned().unwrap_or_default();
        next.insert("_truncated".into(), Value::Array(kept));
        let next = Value::Object(next);
        if approx_tokens(&next) >= approx_tokens(out) {
            return false;
        }
        applied = merged;
        *out = next;
        true
    };

    let mut limit = MAX_ROWS;
    while limit >= 1 && approx_tokens(&out) > max_tokens {
        let mut next = Map::new();
        for (k, v) in out.as_object().unwrap() {
            next.insert(
                k.clone(),
                match v.as_array() {
                    Some(a) => Value::Array(a.iter().take(limit).cloned().collect()),
                    None => v.clone(),
                },
            );
        }
        stage(Value::Object(next), "arrays", format!("→ {limit} row(s)"), &mut out);
        limit -= 1;
    }

    // `objectDepth(out)` is the JS for-INIT — evaluated ONCE, against the payload as it stood when
    // the loop began, not re-measured after each prune.
    let mut level = object_depth(&out).saturating_sub(1);
    while level >= 1 && approx_tokens(&out) > max_tokens {
        let candidate = prune_below(&out, level, 0);
        stage(candidate, "nesting", format!("→ elided below level {level}"), &mut out);
        level -= 1;
    }

    let mut keys = 16usize;
    while keys >= 1 && approx_tokens(&out) > max_tokens {
        let candidate = narrow_to(&out, keys);
        stage(candidate, "width", format!("→ {keys} field(s)"), &mut out);
        keys /= 2;
    }

    if approx_tokens(&out) > max_tokens {
        let candidate = clip_deep(&out, 120);
        stage(candidate, "strings", "→ clipped to 120 chars".to_owned(), &mut out);
    }
    out.as_object().cloned().unwrap_or_default()
}

/// Shape a tool result for the caller's context.
///
/// `verbosity: "full"` returns the payload untouched (the escape hatch for a genuine deep drill).
/// Otherwise: scalars + verdict + the head of each ranked array + a one-line coverage note, under a
/// hard token ceiling, with every truncation disclosed in `_truncated`.
pub fn leanify(result: &Value, verbosity_full: bool, max_tokens: Option<f64>) -> Value {
    if verbosity_full || result.is_null() {
        return result.clone();
    }
    let max_tokens = max_tokens.unwrap_or(DEFAULT_MAX_TOKENS);

    // A pre-rendered text payload ({format,text}) is the tool's OWN compact rendering — only cap it.
    // Re-shaping it would destroy a table the tool deliberately formatted.
    if let Some(o) = result.as_object() {
        if let Some(text) = o.get("text").and_then(Value::as_str) {
            let budget = (max_tokens * CHARS_PER_TOKEN) as usize;
            let n = utf16_len(text);
            if n <= budget {
                return result.clone();
            }
            let mut out = o.clone();
            out.insert(
                "text".into(),
                Value::String(format!("{}\n… truncated ({n} chars) — use verbosity:\"full\"", js_slice(text, budget))),
            );
            return Value::Object(out);
        }
        return Value::Object(enforce_ceiling(shape_generic(o), max_tokens));
    }
    if let Some(a) = result.as_array() {
        let (rows, note) = head_of(a, MAX_ROWS, "rows");
        let mut m = Map::new();
        m.insert("rows".into(), Value::Array(rows));
        if let Some(n) = note {
            m.insert("_truncated".into(), Value::Array(vec![Value::String(n)]));
        }
        return Value::Object(enforce_ceiling(m, max_tokens));
    }
    result.clone()
}
